File: backend/app/core/tick.py
"""Aturan tick size IDX. SPEC.md Bagian 8.8.

Pure function, pakai Decimal (CLAUDE.md: "Semua harga: Decimal, jangan
float"). Semua harga output signal engine WAJIB lewat `round_to_tick()`.
"""

from __future__ import annotations

from decimal import ROUND_CEILING, ROUND_DOWN, ROUND_HALF_UP, Decimal
from typing import Literal

# SPEC.md Bagian 8.8: (batas_bawah_inklusif, batas_atas_eksklusif, tick_size)
TICK_TABLE: list[tuple[Decimal, Decimal | None, Decimal]] = [
    (Decimal(0), Decimal(200), Decimal(1)),
    (Decimal(200), Decimal(500), Decimal(2)),
    (Decimal(500), Decimal(2000), Decimal(5)),
    (Decimal(2000), Decimal(5000), Decimal(10)),
    (Decimal(5000), None, Decimal(25)),
]

RoundDirection = Literal["nearest", "up", "down"]


def get_tick_size(price: Decimal) -> Decimal:
    for low, high, tick in TICK_TABLE:
        if price >= low and (high is None or price < high):
            return tick
    return TICK_TABLE[-1][2]


def round_to_tick(price: Decimal, direction: RoundDirection = "nearest") -> Decimal:
    """Bulatkan harga ke kelipatan tick size yang valid di bursa.

    `direction='down'` dipakai untuk stop loss/take profit sell-side supaya
    tidak overshoot; `'up'` untuk batas atas (mis. entry_high); `'nearest'`
    untuk harga referensi/tampilan umum.
    """
    if price <= 0:
        raise ValueError("price harus > 0")

    tick = get_tick_size(price)
    quotient = price / tick

    if direction == "down":
        rounded_quotient = quotient.to_integral_value(rounding=ROUND_DOWN)
    elif direction == "up":
        rounded_quotient = quotient.to_integral_value(rounding=ROUND_CEILING)
    else:
        rounded_quotient = quotient.to_integral_value(rounding=ROUND_HALF_UP)

    return rounded_quotient * tick

File: frontend/lib/tick.ts
/**
 * Aturan tick size IDX. SPEC.md Bagian 8.8.
 * Pure function, sama seperti versi Python di backend — dipakai di sini
 * untuk membulatkan angka mock supaya realistis (kelipatan tick yang valid).
 */

const TICK_TABLE: [number, number | null, number][] = [
  [0, 200, 1],
  [200, 500, 2],
  [500, 2000, 5],
  [2000, 5000, 10],
  [5000, null, 25],
];

export function getTickSize(price: number): number {
  for (const [low, high, tick] of TICK_TABLE) {
    if (price >= low && (high === null || price < high)) return tick;
  }
  return 25;
}

export function roundToTick(
  price: number,
  direction: "nearest" | "up" | "down" = "nearest"
): number {
  const tick = getTickSize(price);
  if (direction === "down") return Math.floor(price / tick) * tick;
  if (direction === "up") return Math.ceil(price / tick) * tick;
  return Math.round(price / tick) * tick;
}
